addon/-private/system/model/internal-model-class.js

```diff
@@ -0,0 +1,615 @@
+  if (type.superclass) {
+    findPossibleInverses(type.superclass, inverseType, name, possibleRelationships);
+  }
+
+  return possibleRelationships;
+}
+
+
+export default class InternalModelClass {
+  constructor(modelClass) {
+    let attributes;
+    let relationships;
+    let descriptors = get(modelClass, '_computedProperties');
+    let relationshipMap;
+    let properties = descriptors.length ? Object.create(null) : null;
+
+    for (let i = 0; i < descriptors.length; i++) {
+      let value = descriptors[i];
+      let key = value.name;
+
+      if (value && value.meta) {
+        let propertyMeta = value.meta;
+        propertyMeta.key = key;
+
+        if (propertyMeta.isAttribute === true) {
+          if (!attributes) {
+            attributes = Object.create(null);
+          }
+
+          propertyMeta.kind = 'attribute';
+          properties[key] = propertyMeta;
+          attributes[key] = propertyMeta;
+
+        } else if (propertyMeta.isRelationship === true) {
+          if (!relationships) {
+            relationships = Object.create(null);
+            relationshipMap = Object.create(null);
+          }
+          let relationshipMeta = relationshipFromMeta(propertyMeta);
+
+          properties[key] = propertyMeta;
+          relationships[key] = relationshipMeta;
+
+          relationshipMap[propertyMeta.type] = relationshipMap[propertyMeta.type] || [];
+          relationshipMap[propertyMeta.type].push(propertyMeta);
+        }
+      }
+    }
+
+
+    /**
+     Represents the model's class name as a string. This can be used to look up the model through
+     DS.Store's modelFor method.
+
+     `modelName` is generated for you by Ember Data. It will be a lowercased, dasherized string.
+     For example:
+
+     ```javascript
+     store.modelFor('post').modelName; // 'post'
+     store.modelFor('blog-post').modelName; // 'blog-post'
+     ```
+
+     The most common place you'll want to access `modelName` is in your serializer's `payloadKeyFromModelName` method. For example, to change payload
+     keys to underscore (instead of dasherized), you might use the following code:
+
+     ```javascript
+     export default const PostSerializer = DS.RESTSerializer.extend({
+     payloadKeyFromModelName: function(modelName) {
+       return Ember.String.underscore(modelName);
+     }
+   });
+     ```
+     @property modelName
+     @type String
+     @readonly
+     @static
+     */
+    this.modelName = modelClass.modelName;
+    this.modelClass = modelClass;
+    this._attributes = attributes;
+    this._relationships = relationships;
+    this._properties = properties;
+
+    // for deprecated things
+    this._relationshipMap = relationshipMap;
+  }
+
+  metaForProperty(name) {
+    return this._properties[name];
+  }
+
+  eachRelationship(cb, context) {
+    let relationships = this._relationships;
+
+    if (relationships) {
+      let exec = cb.bind(context);
+
+      for (let name in relationships) {
```

Comment:
we seem to only iterate relationships for there membership, should `this._relationships` just be an array then?

it will then:

* iterate faster
* be easy to skip/avoid/dealloc if empty